### PR TITLE
Add salesforce common module

### PR DESCRIPTION
**Pull request by david-pepper**
## What does this change?
Add a Salesforce module that contains actions  common across typescript projects in the repo, such as:

- config e.g. current Salesforce api version
- authentication with Salesforce using connected app and api user credentials
- query for records in Salesforce using REST api
- update records in Salesforce using REST api

test modules have been added for each module

modules have been plugged into from the new zuora-salesforce-link-remover project

This module should be used or added to for Salesforce integrations in future, and retrofitted to existing solutions when it makes sense to do so

### CODE REVIEW

File: handlers/zuora-salesforce-link-remover/src/handlers/getBillingAccounts.ts [hunk 1]
```diff
@@ -36,16 +35,35 @@ export async function handler() {
...
+
+export const BillingAccountRecordSchema = RecordSchema.extend({
+	GDPR_Removal_Attempts__c: z.number(),
+	Zuora__External_Id__c: z.string(),
+});
+export type BillingAccountRecord = z.infer<typeof BillingAccountRecordSchema>;
+
+//might need hoisting
+export const BillingAccountRecordsSchema = z.array(BillingAccountRecordSchema);
+export type BillingAccountRecords = z.infer<typeof BillingAccountRecordsSchema>;
```
Comment (tomrf1) [inline handlers/zuora-salesforce-link-remover/src/handlers/getBillingAccounts.ts:61]: is this type useful? I think I'd find `BillingAccountRecord[]` more readable

File: modules/salesforce/src/query.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+				},
+			},
+		);
+
+		if (!response.ok) {
+			throw new Error(`Failed to execute query: ${response.statusText}`);
+		}
+
+		const sfQueryResponse =
+			(await response.json()) as SalesforceQueryResponse<T>;
```
Comment (tomrf1) [inline modules/salesforce/src/query.ts:29]: I think the cast here is unnecessary if you're parsing it immediately after

File: modules/salesforce/src/query.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import type { SfAuthResponse } from './auth';
+import { sfApiVersion } from './config';
+import type { SalesforceQueryResponse } from './recordSchema';
+import { SalesforceQueryResponseSchema } from './recordSchema';
+
+export async function executeSalesforceQuery<T extends z.ZodTypeAny>(
+	sfAuthResponse: SfAuthResponse,
+	query: string,
+	schema: T,
+): Promise<SalesforceQueryResponse<z.infer<T>>> {
```
Comment (tomrf1) [inline modules/salesforce/src/query.ts:11]: is `z.infer` necessary here?

**Comment by tomrf1** [comment]
I found the docs [here](https://github.com/colinhacks/zod?tab=readme-ov-file#writing-generic-functions) useful for following this PR

File: modules/salesforce/src/query.ts (hunk 2 above)
Comment (david-pepper) [reply modules/salesforce/src/query.ts:29]: yep, you're right. I've removed the casting.

File: modules/salesforce/src/query.ts (hunk 3 above)
Comment (david-pepper) [reply modules/salesforce/src/query.ts:11]: Nope. Removed it

File: handlers/zuora-salesforce-link-remover/src/handlers/getBillingAccounts.ts (hunk 1 above)
Comment (david-pepper) [reply handlers/zuora-salesforce-link-remover/src/handlers/getBillingAccounts.ts:61]: agreed, I've updated the code to use ```z.array(BillingAccountRecordSchema)``` instead

File: modules/salesforce/src/config.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+export const sfApiVersion = (): string => {
+	const sfApiVersion = process.env.SF_API_VERSION;
+
+	if (!sfApiVersion) {
+		return 'v54.0';
```
Comment (charleycampbell) [inline modules/salesforce/src/config.ts:5]: Does this version change frequently? I wonder if a code comment explaining why v54.0 is used if sfApiVersion is false

**Review by charleycampbell — APPROVED** [review]
I really like your console logging through out your PR - it's really been considered where to put it and will help if we ever need to trawl through the logs!

File: modules/salesforce/src/config.ts (hunk 4 above)
Comment (david-pepper) [reply modules/salesforce/src/config.ts:5]: Good shout!

**Pull request merged**